text/3373-avoid-nonlocal-definitions-in-fns.md

```diff
@@ -0,0 +1,71 @@
+- Feature Name: N/A
+- Start Date: 2022-01-19
+- RFC PR: [rust-lang/rfcs#3373](https://github.com/rust-lang/rfcs/pull/3373)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Starting in Rust 2024, stop allowing items inside functions or expressions to
+implement methods or traits that are visible outside the function or
+expression.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, tools cross-referencing uses and definitions (such as IDEs) must
+either search inside all function bodies and other expression-containing items
+to find potential definitions corresponding to uses within another function, or
+not cross-reference those definitions at all.
+
+Humans cross-referencing such uses and definitions may find themselves
+similarly baffled.
+
+With this change, both humans and tools can limit the scope of their search and
+avoid looking for definitions inside other functions or items, without missing
+any relevant definitions.
+
+# Explanation
+[explanation]: #explanation
+
+The following types of items, "expression-containing items", can contain
+expressions, including the definitions of other items:
+- Functions
+- Closures
+- The values assigned to `static`/`const` items
+- The discriminant values assigned to `enum` variants
+
+Starting in the Rust 2024 edition:
+- An item nested inside an expression-containing item (through any level of
+  nesting) may not define an `impl Type` block unless the `Type` is also nested
+  inside the same expression-containing item.
+- An item nested inside an expression-containing item (through any level of
+  nesting) may not define an `impl Trait for Type` unless either the `Trait` or
+  the `Type` is also nested inside the same expression-containing item.
+- An item nested inside an expression-containing item (through any level of
+  nesting) may not define an exported macro visible outside the
+  expression-containing item (e.g. using `#[macro_export]`).
```

Comment:
We forgot to mention modules and includes (thanks @matklad for the reminder)

```suggestion
  expression-containing item (e.g. using `#[macro_export]`).
- An item nested inside an expression-containing item (through any level of
  nesting) may not use mod declarations referring to other files 
  (`#[path = "foo.rs"] mod foo;`) or `include!()` macro.
```

Reply:
What would be the rustfixed version of this? Any module declared in a function is not reachable from the outside, so you can't just move the declaration outside of the function. `include` is even harder as it can be in any part of the code. You can paste the entire contents of the file but it might be autogenerated and you don't want it inside that file. E.g. autogeneration plus usage inside a fn body is done in [`rust_codegen_gcc`](https://github.com/rust-lang/rustc_codegen_gcc/blob/8d0ff6bf0776d421ee2a38c09e6d60a8b361c99e/src/intrinsic/llvm.rs#L773).

Reply:
> E.g. autogeneration plus usage inside a fn body is done in [rust_codegen_gcc](https://github.com/rust-lang/rustc_codegen_gcc/blob/8d0ff6bf0776d421ee2a38c09e6d60a8b361c99e/src/intrinsic/llvm.rs#L773).

That `include!()` produces an expression without any nested items, mod declarations or `include!()`.

Reply:
@bjorn3 but it's an `include!` nontheless, and @vlad20012 's proposal is to ban those inside fn bodies as well.

@matklad 's [point](https://github.com/rust-lang/rfcs/pull/3373#issuecomment-1406888432) is that they want to establish where a given .rs file sits in the module hierarchy without macro-expanding inside fn bodies in the rest of the crate, at least that's how I understand it.

Even if you copy-paste what the `include` expands to, as the rustfix suggestion, you'd have to manually adjust your autogeneration script to export something that can live outside of a fn body (like a function). It might also not be checked into git, glium has an `include` of a file that is being generated in build.rs (admittedly it's not being expanded in an fn context though).

Reply:
I think `include` is more like macro bodies and less like modules since it can be used in multiple places, and autocomplete inside them should work without context knowledge. And generally isn't banning include in function body too radical?

Reply:
> I think `include` is more like macro bodies and less like modules since it can be used in multiple places, and autocomplete inside them should work without context knowledge. And generally isn't banning include in function body too radical?

I've gotten a bit lost, why is `include!` being considered for banning withing function bodies?

Reply:
From https://github.com/rust-lang/rfcs/pull/3373#issuecomment-1406888432
> A case which is *probably* wrong is
> ```rust
> fn foo() {
>    #[path = "/my/file.rs"]
>    mod foo;
> }
> ```
> If we do coherence thing, this'll actually be OK for parallel compilation. However, for IDEs there will be a bad case of providing completion inside `/my/file.rs`, as IDE won't know where in the module hierarchy that file is mounted without looking inside `foo`.

Reply:
I see. I see how they are similar but imho they are not the same. For example I could just be including an HTML template and might not be including a module. If we restrict compilation to disallow the things inside the function to have global effect I don't see this as a major issue. IDE performance would only be affected when working inside of that function the rest of the time it should be able to be ignored.

Reply:
Agreed, for determining the rs file's position in the module hierarchy, one could have an algorithm that first checks for all mod statements *outside* of function bodies. Then, if nothing is found, you can macro expand inside function bodies in a second pass. So users would only be affected if they are inside a file that was included from a function body, which should be rare.

The `impl` that is accessible from outside question is different, because due to how Rust's resolution code works, an impl in a function body has priority over a similarly named function in an `impl Trait` body at top level. So without this RFC, you always have to visit function bodies, even if there is nothing inside them, while for `mod` statements that's only needed as a fallback.

Reply:
> while for mod statements that's only needed as a fallback.

Not really, a file can be included into module hierarchy more than once. The fact that you’ve found a march outside doesn’t mean that there’s no match inside as well.

Reply:
Fair. I've thought that IDEs don't care much for that case because getting multiple locations from the question "where does this file stand in the module hierarchy?" makes the following algorithms more complicated as they have to deal with multiple locations, so having *one* file even if it's not the only one is enough for IDEs.

Personally I've never seen the same file being included multiple times in the same compilation unit. It might occur with some macros though. The only thing I *have* seen is multiple compilation units all using the same file, but usually that uses `#[path]` (tests or examples all using an utils module). The suggested change would entirely disable one of the features of include, being used in expression context.

What about forbidding *mixed* uses? As in, either it's used always inside an `fn` body, or it is never used inside an `fn` body.

Reply:
I don't really understand why include needs to be forbidden. How does that help?

Reply:
@c-git it's been explained above, but to expand on it, consider e.g. (in different files):

```Rust
mod foo { // foo.rs
    struct Hi;
    mod bar { // foo/bar.rs
        struct Hi;
        #[path = "../baz.rs"]
        mod baz;
    }
}
```

And then in `foo/baz.rs` you have a statement: `use super::Hi;`.

If you open baz.rs in your IDE, which `Hi` struct does it resolve to? `foo::Hi` or `foo::bar::Hi`? Is it maybe a resolution error that the user needs to be told about? You only know that if you can figure out baz's position in the module hierarchy, for which you need to follow the chain of `mod foo` statements from the top-level `foo.rs` file to `foo/bar.rs` to `foo/baz.rs`. You find the top level file by re-implementing cargo's Cargo.toml search algorithm, then deriving it from there, or by being passed that top level file's path on the CLI when another build system is involved.

We are assuming an IDE here that is query-based and that tries to minimize the work to answer that query. If include (or `#[path = ...] mod baz;`) are forbidden, then that means that it doesn't have to parse the bodies of functions to answer queries. Minimizing the needed work means the IDE is fast and snappy instead of slow, which is what everyone wants :).

So yes, there are arguments in favour of disallowing `include` in fn bodies. But personally, I think that for non-local impls in functions there is a way stronger case to be made for disallowing it:

* it is rustfix-able (while include and mod statements are not, as discussed above)
* already now, it's quite weird to have `impl`s in function bodies for outside types. I wouldn't call an `include` in a fn body really weird.
* supporting non-local `impl` blocks in functions means you have to do more computations, you have to resolve the impl block's type name at least, and then add it to a list of impl blocks the type, for all impl blocks in the crate, nested or not. Then, you'd to do the fine-grained parsing for the type you are interested in. usually you don't just want to resolve stuff for one thingy alone, you'd want to resolve everything and create a new IR for it, so that creates additional overhead. For mod statements on the other hand, you don't need to go to full resolution, all you need is the post-expansion AST (in rustc compiler terms), and that would only be the fallback to when a file is not found in the hierarchy. the non-fallback would just work on the `TokenTree` representation for the fn bodies, and post-expansion AST for the top level stuff.

Reply:
I think that disallowing `include!` in function bodies should be a separate RFC.

Partly because `include!`'s current behaviour *in the context of the compiler* isn't as big a confusing footgun as non-local definitions in functions (Just have the compiler allow or deny `include!`s based on whether their contents are valid in that context, same as what would go on with a C preprocessor `#include`) and partly because I'm concerned that aversion to disallowing `include!` in function bodies might be "the rider that gets the desirable bill killed".

(i.e. I worry that, if the push is for "change both or neither", we'll get "neither"... and I'm on the fence about whether I'd consider it problematic enough to push for "change nothing" to avoid losing the option of using `include!` inside functions.)

Reply:
Thank you for explaining.

Reply:
> @est31 Personally I've never seen the same file being included multiple times in the same compilation unit.

I did this a few times. Can't say whether it was a good idea, though. The context was a test suite which was repeated almost verbatim for several similar but meaningfully different types implementing the same interfaces. I could just duplicate the test modules manually, which was no fun, since it contained half a hundred functions. I could extract the test functions into generic functions - but that would need huge trait bounds, and when a test is 1-2 function calls anyway there is no win from generics. I could write a macro which generated the entire module - but that would break name resolution, formatting, IDE navigation and test runner, and just wouldn't be that readable.

So instead I wrote a single test file which ran some tests on the name `TestType`, and each tested module would have this block:
```rust
#[cfg(test)]
mod tests {
    use super::*;
    use super::Foo as TestType;
    include!("path/to/test.rs");
}
```
Alternatively, if the tested structs were named the same, I could move that prelude into the test suite file, and do
```rust
#[cfg(test)]
#[path("path/to/test.rs)"
mod tests;
```
The IDE handled this multiple include a bit incoherently, with a single file being selected as its supermodule, but otherwise everything worked as expected: syntax highlighting, formatting, name resolution, type inference, test discovery. This pattern is applicable only to niche cases, and it's a bit obscure, so I'd think twice about using it, but in all other ways I'm quite happy with the results.

Reply:
@afetisov interesting!

> The IDE handled this multiple include a bit incoherently, with a single file being selected as its supermodule, but otherwise everything worked as expected

Yeah this means that the IDE is currently only looking for the first occurrence in the hierarchy, and then exiting. Which gives you most of the features. If you want to do it perfectly, then the algorithms that take in the data [become more complicated](https://github.com/rust-lang/rfcs/pull/3373#discussion_r1089817482). Like when you use the "where is item X used" feature, if the item you click on is actually two items because the file is included twice, you have to do the search for both items separately, then merge the lists. For autocompletion inside the file you need to do the opposite, to consider the intersection.

----------

In addition to my suggestion from above to forbid mixed as-item and not-as-item uses as an alternative to forbid not-as-item uses of `include`, one could also add an `allow_not_unique` annotation to `include`, e.g. via `include!("path/to/file.rs", allow_not_unique)`.

For IDEs this would mean that they can terminate their search at the first inclusion if they don't find an `allow_not_unique` and continue it if they find one. This would roughly halve the search time on average, on top of the improvements for not going inside items. For rustc it would mean that it errors if, in a new edition, a file is included multiple times and there is no such annotation. That annotation would also work on `#[path]` as another attribute. The annotation would also make the error rustfix-able, which was my original concern at the top of this sub-thread.

Regarding `include` inside items, the search for that would also only be triggered if there is no `allow_not_unique`-free `include` present. 

The only problem I see is that people might add it too often, as `allow_not_unique` would still work on a per-crate level, not on a per-package level, so if multiple files in the `examples/` dir all `include` a file they wouldn't need it. This actually causes a bit of a tension between an IDE and rustc, because from the POV of the IDE, it would make more sense for it to work on a per-package level than on a per-crate level, as a file could belong to any compilation unit(s). But the compiler can't really think beyond its current compilation unit.

Reply:
Considering how controversial the forbidding of `include` has turned out to be, I think we'd better not try to bring it into this RFC. I don't see the reason to forbid mod declarations as well if includes will still be allowed.